doc/specs/stdlib_hash_procedures.md

```diff
@@ -0,0 +1,1748 @@
+
+##### Syntax
+
+`code = [[stdlib_32_bit_hash_codes:water_hash]]( key, seed )`
+
+##### Class
+
+Pure/elemental function
+
+##### Arguments
+
+`key`: Shall be a deferred length default character scalar expression
+or a rank 1 integer array expression of kind `INT8`, `INT16`,
+`INT32`, or `INT64`.
+It is an `intent(in)` argument.
+
+`seed`: shall be an integer scalar expression of kind `INT64`.
+It is an `intent(in)` argument.
+
+##### Result
+
+The result is a scalar integer of kind `INT32`.
+
+##### Note
+
+`WATER_HASH` is an implementation of the `waterhash` hash code of
+Tommy Ettinger.
+This code has excellent performance on long keys, and good performance
+on short keys.
+As a result it should give reasonable performance for typical hash
+table applications.
+This code passes the SMHasher tests.
+The `waterhash` is based on the `wyhash` of Wang Yi.
+While `wyhash` has a number of bad seeds, where randomization of the
+output is poor,
+so far testing has not found any bad seeds for `waterhash`.
+It can have undefined behavior if the key is not word aligned,
+i.e. some computer processors can only process a given size integer if
+the address of the integer is a multiple of the integer size. 
+It is a *pure* function for integer arrays, and an *elemental*
+function for character strings. 
+
+##### Example
+
+```fortran
+    program demo_water_hash
+      use stdlib_32_bit_hash_codes, only: water_hash, &
+	  new_water_hash_seed
+      use iso_fortran_env, only: int32, int64
+      implicit none
+      integer(int32) :: hash
+      integer(int64) :: seed = 42_int64
+      call new_water_hash_seed( seed )
+      hash = water_hash([ 5, 4, 3, 1, 10, 4, 9], seed)
+      print *, hash, seed
+    end program demo_water_hash
+```
+
+## The `stdlib_64_bit_hash_codes` module
+
+### Overview of the module
+
+Sixty four bit hash functions are generally overkill for hash table
+applications, and are primarily useful for check sums and related
+applications.
+As checksums often have to deal with extremely large files or
+directories, it is often useful to use incremental hashing as well as
+direct hashing, so 64 bit and higher hash algorithms often provide
+multiple implementations. The current module, for simplicity of API,
+doesn't provide any incremental hashes.
+The `stdlib_64_bit_hash_codes` module defines several public
+overloaded 64 bit hash procedures, `FNV_1`, `FNV-1A`,
+`PENGY_HASH`, and `SPOOKY_HASH`, two scalar hash functions,
+`FIBONACCI_HASH` and 
+`UNIVERSAL_MULT_HASH`, a seed generator, `ODD_RANDOM_INTEGER`, for the
+`UNIVERSAL_MULT_HASH`, and two seed generators, `NEW_PENGY_HASH_SEED`
+and `NEW_SPOOKY_HASH_SEED` for their respective hash functions. It
+also defines the integer kind constant, `INT_HASH`, used to specify
+the kind of the hash function results, and a logical constant,
+`LITTLE_ENDIAN`, used to deal with one aspect of the machine
+dependence of the hash codes. 
+Note that while SpookyHash can be used as a sixty four bit hash
```

Comment:
```suggestion
Note that while SpookyHash can be used as a sixty-four bit hash
```

I've noticed in some places the forms 64 bit and 128 bit are used. I believe these should be hyphenated too.

Reply:
Do you want the hyphen between the "64" and the "bit"?